staging/src/k8s.io/kubelet/config/v1beta1/types.go

```diff
@@ -677,6 +677,10 @@ type KubeletConfiguration struct {
 	// Default: ""
 	// +optional
 	ProviderID string `json:"providerID,omitempty"`
+	// ExternalCloudProvider replaces --cloud-provider=external.  When set to
+	// true, an external cloud provider is expected to be used.
+	// +optional
+	ExternalCloudProvider *bool `json:"externalCloudProvider,omitempty"`
```

Comment:
I'm not sure if there will be any other features in this provider beyond what `registerWithTaints` can offer going forward. I also see some node-ip related annotation, but it feels that may not be needed in a long run.

Do you envision more features added to the external cloud provider behavior?

Reply:
the node-ip is a problematic one that also has a KEP in flight https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3705-cloud-node-ips

/cc @danwinship

Reply:
> Do you envision more features added to the external cloud provider behavior?

Good question. At this point, no; but I think others would have more context than me.

Applying `node.cloudprovider.kubernetes.io/uninitialized` "by hand" is a bit clunkier than this option, but would suffice.

Reply:
> I also see some node-ip related annotation, but it feels that may not be needed in a long run.

It's not just the annotation; when an external cloud provider is present, kubelet expects the cloud provider to manage `.status.addresses` on the Node (and it sets the annotation to help it do that), whereas when there is no external cloud provider, kubelet manages that field itself. If kubelet didn't know whether or not there was a cloud provider then it would end up fighting with the cloud provider over what the correct value of that field was.

Reply:
@danwinship is correct: https://github.com/kubernetes/kubernetes/blob/64259b43b8e5b7c086bf2b74743dc7e73ebe37e3/pkg/kubelet/nodestatus/setters.go#L96-L112

I think `KubeletConfiguration.ExternalCloudProvider` is preferable to users having to handle these details.

Reply:
Will this https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3705-cloud-node-ips fix the IP portion of it? I didn't read in details, but I see there will be another field added there as well. Will it still need this bit (external or not) present on config?

Reply:
That KEP changes some of the specific details around node IP handling, but it doesn't change the overall architecture, where kubelet behaves differently with an external cloud provider than it does with no cloud provider (or a legacy cloud provider).

(Theoretically, this could be changed if there's a strong desire to make kubelet not need to explicitly know whether there's a cloud provider. Rather than having the cloud-provider manage `node.status.addresses` directly itself, it could just set an annotation on the Node and then kubelet would read that annotation and set `node.status.address` accordingly. There would be all sorts of upgrade and skew considerations so this would need a KEP to figure out the details.)

Reply:
> Is there scenario when one may want address managed externally but start without taints?

Right now it's guaranteed that pods won't get scheduled to the node until the cloud-provider has figured out the node IPs. (Because the node initially comes up tainted, and the cloud-provider doesn't remove the taint if it can't figure out the IPs; eg, if the node is not actually known to the cloud provider.)

So unless there was some larger refactoring, then no, you wouldn't want external-node-IP-management without having the taint as well.